extensions/apns_message_handler.go

```diff
@@ -272,87 +242,64 @@ func (a *APNSMessageHandler) buildNotification(message interfaces.KafkaMessage)
 	return notification, nil
 }
 
-func (a *APNSMessageHandler) sendNotification(notification *Notification) error {
-	l := a.Logger.WithField("method", "sendNotification")
+func (a *APNSMessageHandler) buildAndValidateNotification(notification *pusherAPNSKafkaMessage) (*structs.ApnsNotification, error) {
 	if notification.PushExpiry > 0 && notification.PushExpiry < MakeTimestamp() {
-		l.Warnf("ignoring push message because it has expired: %s", notification.Payload)
-
-		apnsResMutex.Lock()
-		a.ignoredMessages++
-		apnsResMutex.Unlock()
-
-		if a.pendingMessagesWG != nil {
-			a.pendingMessagesWG.Done()
-		}
-		return errors.New("expired message")
+		return nil, errors.New("push message has expired")
 	}
+
 	payload, err := json.Marshal(notification.Payload)
 	if err != nil {
-		l.WithError(err).Error("error marshaling message payload")
-
-		apnsResMutex.Lock()
-		a.ignoredMessages++
-		apnsResMutex.Unlock()
+		return nil, fmt.Errorf("error marshalling notification payload: %s", err.Error())
+	}
 
-		if a.pendingMessagesWG != nil {
-			a.pendingMessagesWG.Done()
-		}
-		return err
+	n := &structs.ApnsNotification{
+		Notification: apns2.Notification{
+			Topic:       a.ApnsTopic,
+			DeviceToken: notification.DeviceToken,
+			Payload:     payload,
+			ApnsID:      notification.ApnsID,
+			CollapseID:  notification.CollapseID,
+		},
+		Metadata: notification.Metadata,
 	}
-	l.WithField("notification", notification).Debug("adding notification to apns push queue")
+	return n, nil
+}
+
+func (a *APNSMessageHandler) sendNotification(notification *structs.ApnsNotification) {
 	before := time.Now()
 	defer statsReporterReportSendNotificationLatency(a.StatsReporters, time.Since(before), a.appName, "apns", "client", "apns")
-	a.PushQueue.Push(&apns2.Notification{
-		Topic:       a.ApnsTopic,
-		DeviceToken: notification.DeviceToken,
-		Payload:     payload,
-		ApnsID:      notification.ApnsID,
-		CollapseID:  notification.CollapseID,
-	})
-	return nil
+
+	notification.SendAttempts += 1
```

Comment:
why don't we increment this closer to where we're sending? for example when we read from the push queue and call APNS

Reply:
I think conceptually it makes more sense for the handler to control how many times a message is sent. The handler is where we retry and where we check if the amount of attemps is whithin the limit. If we put it in the client we would be spliting this responsibility in a way.

Reply:
Do you think it makes sense?

Reply:
Makes sense, just the weird achitecture that would make it possible for the handler to increment but this message never actually gets attempted because it was not read from the pushQueue channel